### ISSUE TITLE
How to build and deploy all those built files to a hosting server (ex: firebase) without server-side nodejs environment? 

### ISSUE THREAD

**Issue by EasonPai** [react-boilerplate/react-boilerplate#2243]
## Description

Is it possible to build and deploy all those built files to a hosting server (ex: firebase) without server-side nodejs environment? 

I can successfully 'npm run start' on my local machine, everything works perfect.
but when I upload built files to hosting server and navigate to url like: 'https://xxxxx.firebaseapp.com' . I got error messages below:

> https://xxxxx.firebaseapp.com/main.a68767237fa35b2c54de.js 
> Status Code: 404  (from disk cache)

if I navigate to url like: 'https://xxxxx.firebaseapp.com/index.html',
 I can see my HomePage, but the router shows "Page not found."

<img width="826" alt="pagenotfound" src="https://user-images.githubusercontent.com/678352/41330698-a933d11e-6f07-11e8-926d-4355539065ef.png">

What is the correct workflow to deploy built files to a hosting server?

# Not a SPA?
Besides, is it possible to make my app simply a webpage instead of a website?
For example, I want to provide a new webpage and host it at someplace like:

> https://www.example.com/new/version/under/this/path

Then I can still have my original web app running at `https://www.example.com`.

This is what I really want, but I guess I have to make it deploy as a SPA first,
then step to next stage to make it a webpage.

# Versions

- React-Boilerplate (see `package.json`): 3.5.0
- Node/NPM: 5.0.3
- Browser: Chrome

**Comment by julienben** [comment]
You can deploy your apps to Firebase Hosting without any server-side code at all. It's all taken care of on their side. You just need to make sure that it treats your app as an SPA (by redirecting all requests to your index.html file). This is done with the following `firebase.json` config:

```json
{
  "hosting": {
    "public": "build",
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ]
  }
}
```
I suggest following [their guide](https://firebase.google.com/docs/hosting/) directly...

As for your second question... You mean you're trying to deploy a React/RBP app directly to a subfolder of a static website you already have? Yes that should be possible as well. Maybe even with Firebase (although you'll have to dig into their documentation to find out). 

Basically, you want something that'll automatically serve your static site on the relevant URLs but, if any requests come through for your SPA path (https://www.example.com/new/version/under/this/path) or any lower paths (i.e. paths that begin with https://www.example.com/new/version/under/this/path), you'll want to server your app's index.html instead. You'll also need to change the client-side router so that it's aware of its sub-path.

Not easy but totally feasible. Good luck ;)

**Comment by EasonPai** [comment]
Thanks for point out the need of hosting configurations, it really helps for beginners like me to clarify some important basics.

But I still have some problem to make it work. 

### For Hosting settings
I follow firebase official docs and set my firebase.json, I tried to 'rewrites' my path something like:
```
{
        "source": "**/newpath/",
        "destination": "/newpath/index.html"
}
```

### For React-Boilerplate settings
In order to let my react-boilerplate SPA knows where to load all built files, I revised 'publicPath' from `'/'` to  `'./'`. 

### Result

After I deploy built files to firebase and navigate to `https://www.example.com/newpath`, everything works fine, 
- `https://www.example.com/newpath` leads to my SPA index.html 
- `https://www.example.com/newpath/main.xxxx.js` loads the right js

only Chrome said it cannot find 
- `https://www.example.com/sw.js` 
  > Referrer Policy: no-referrer-when-downgrade

This is strange, is't it supposed to load `sw.js` from `https://www.example.com/newpath/sw.js`? 
I tried to play around with settings of OfflinePlugin, but all lead to the same result.

What should I do then?

**Comment by julienben** [comment]
OfflinePlugin is only in the dev branch. Did you use that? If so, look [here](https://github.com/NekR/offline-plugin/blob/master/docs/options.md) and be very thorough about understanding the options because you have a unique setup and you might need to teach the service worker how to understand it. If you can't get it to work, I would just get rid of it if I were you.

**Comment by julienben** [comment]
I'm gonna lock this conversation as it's not relevant/specific to RBP anymore.
